sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -0,0 +1,193 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.types._
+
+/**
+ * This aims to handle nested column aliasing pattern inside `ColumnPruning` optimizer rule.
+ * If a project or its child references to nested fields, and not all the fields
+ * in a nested attribute are used, we can substitute the by alias attributes; then a project
+ * of the nested fields as aliases on the children of the child will be created.
+ */
+object NestedColumnAliasing {
+
+  def unapply(plan: LogicalPlan)
+    : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = plan match {
+    case Project(_, child) if canProjectPushThrough(child) =>
+      getAliasSubMap(plan, child)
+    case _ => None
+  }
+
+  /**
+   * Replace nested columns to prune unused nested columns later.
+   */
+  def replaceToAliases(
+      plan: LogicalPlan,
+      nestedFieldToAlias: Map[GetStructField, Alias],
+      attrToAliases: Map[ExprId, Seq[Alias]]): LogicalPlan = plan match {
+    case Project(projectList, g @ GlobalLimit(_, grandChild: LocalLimit)) =>
```

Comment:
We need this match? IIUC, if this match doesn't exist, this rule generates unnecessary aliases between local/global limits. We cannot remove these unnecessary aliases in another optimizer rule, e.g., RemoveNoopOperators?

I think this rule could generate unnecessary aliases in other cases, too, e.g.,
```

scala> spark.table("t").printSchema
root
 |-- c0: long (nullable = true)
 |-- c1: struct (nullable = true)
 |    |-- n0: long (nullable = true)
 |    |-- n1: long (nullable = true)

scala> spark.table("t").sample(0.5).repartition(1).select("c1.n0").explain(true)
== Optimized Logical Plan ==
Project [_gen_alias_36#36L AS n0#35L]
+- Repartition 1, true
   +- Project [_gen_alias_37#37L AS _gen_alias_36#36L]
      +- Sample 0.0, 0.5, false, 5378592115661057580
         +- Project [c1#25.n0 AS _gen_alias_37#37L]
            +- Relation[c1#25] parquet
```

Reply:
Yes. Right. Current, I didn't add additional optimizer rule to handle that redundant one.

Reply:
Yea, I know the great metric against the additional cost (unnecessary aliases)! So, I just wanted to know which one we should cover in this pr, and which one we could resolve in feature work. If we have TODO tasks, IMO it would be better to leave TODO comments where possible.

Reply:
Sure. Thanks. Noop Alias issue is a general one because users can make it, too. I'll try to file a general JIRA issue for removing Noop Alias.

Reply:
@maropu . Since that is a general issue, I created [SPARK-27123](https://issues.apache.org/jira/browse/SPARK-27123) and make a [PR](https://github.com/apache/spark/pull/24049) for that.